os/drivers/audio/syu645b.c

```diff
@@ -0,0 +1,1132 @@
+}
+
+/****************************************************************************
+ * Name: syu645b_configure
+ *
+ * Description:
+ *   Configure the audio device for the specified  mode of operation.
+ *
+ ****************************************************************************/
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+static int syu645b_configure(FAR struct audio_lowerhalf_s *dev, FAR void *session, FAR const struct audio_caps_s *caps)
+#else
+static int syu645b_configure(FAR struct audio_lowerhalf_s *dev, FAR const struct audio_caps_s *caps)
+#endif
+{
+#if !defined(CONFIG_AUDIO_EXCLUDE_VOLUME) || !defined(CONFIG_AUDIO_EXCLUDE_TONE)
+	FAR struct syu645b_dev_s *priv = (FAR struct syu645b_dev_s *)dev;
+#endif
+	int ret = OK;
+
+	DEBUGASSERT(priv && caps);
+	audvdbg("ac_type: %d\n", caps->ac_type);
+
+	/* SYU645B supports on the fly changes for almost all changes
+	   so no need to do anything. But if any issue, worth looking here */
+
+	switch (caps->ac_type) {
+	case AUDIO_TYPE_FEATURE:
+		audvdbg("  AUDIO_TYPE_FEATURE\n");
+
+		/* Inner swich case: Process based on Feature Unit */
+		switch (caps->ac_format.hw) {
+#ifndef CONFIG_AUDIO_EXCLUDE_VOLUME
+		case AUDIO_FU_VOLUME: {
+			/* Set the volume */
+			uint16_t volume = caps->ac_controls.hw[0];
+			audvdbg("    Volume: %d\n", volume);
+			if (volume < SYU645B_SPK_VOL_MIN) {
+				priv->volume = SYU645B_SPK_VOL_MIN;
+			} else if (volume >= SYU645B_SPK_VOL_MAX) {
+				priv->volume = SYU645B_SPK_VOL_MAX;
+			} else {
+				priv->volume = volume;
+			}
+			printf("set volume is called\n");
+			syu645b_setvolume(priv);
+		}
+		break;
+		case AUDIO_FU_MUTE: {
+			/* Mute or unmute:	true(1) or false(0) */
+			bool mute = caps->ac_controls.b[0];
+			audvdbg("mute: 0x%x\n", mute);
+			priv->mute = mute;
+			syu645b_setvolume(priv);
+		}
+		break;
+#endif							/* CONFIG_AUDIO_EXCLUDE_VOLUME */
+		default:
+			auddbg("    ERROR: Unrecognized feature unit\n");
+			ret = -ENOSYS;
+			break;
+		}
+		break;					/* Break for inner switch case */
+	case AUDIO_TYPE_OUTPUT:
+		audvdbg("  AUDIO_TYPE :%s\n", caps->ac_type == AUDIO_TYPE_INPUT ? "INPUT" : "OUTPUT");
+		/* Verify that all of the requested values are supported */
+		ret = -EDOM;
+		if (caps->ac_channels != SYU645B_CHANNELS) {
+			auddbg("ERROR: Unsupported number of channels: %d\n", caps->ac_channels);
+			break;
+		}
+
+		if (caps->ac_controls.b[2] != SYU645B_BPSAMP) {
+			auddbg("ERROR: Unsupported bits per sample: %d\n", caps->ac_controls.b[2]);
+			break;
+		}
+
+		/* Save the current stream configuration */
+		priv->samprate = caps->ac_controls.hw[0];
+		priv->nchannels = caps->ac_channels;
+		priv->bpsamp = caps->ac_controls.b[2];
+
+		audvdbg("    Number of channels: 0x%x\n", priv->nchannels);
+		audvdbg("    Sample rate:        0x%x\n", priv->samprate);
+		audvdbg("    Sample width:       0x%x\n", priv->bpsamp);
+
+		/* Reconfigure the FLL to support the resulting number or channels,
+		 * bits per sample, and bitrate.
+		 */
```

Comment:
here, we will have to add changes to set i2s and device to the required sampling rate and data width.